File: common/programmers/양궁대회/양궁대회_고진석.java
class 양궁대회 {
	int[] minArr = new int[11];
	boolean[] winPos = new boolean[11];
	int[] answer;
	int max = -100;
	
	public int[] solution(int n, int[] info) {
		for(int i = 0 ; i < 11 ; i++)
			minArr[i] = info[i] + 1;
    	
		selectPos(0, 0, n, info);
		
		if(max <= 0)
			return new int[]{-1};
        return answer;
    }
	
	public void selectPos(int depth, int start, int n, int[] info) {
		if(depth > 0) {
			int[] info1 = new int[11];
			int cnt = 0;
			for(int i = 0 ; i < 10 ; i++)
				if(winPos[i]) {
					info1[i] = minArr[i];
					cnt += info1[i];
				}
			if(cnt > n)
				return;
			info1[10] = n - cnt;
			
			int score = calc(info1, info);
			if(score > max) {
				max = score;
				answer = info1;
			}
			else if(score == max) {
				for(int i = 10 ; i >= 0 ; i--) {
					if(info1[i] > answer[i]) {
						answer = info1;
						break;
					}
					else if(info1[i] < answer[i])
						break;
				}
			}
		}
		for(int i = start ; i < 10 ; i++) {
			winPos[i] = true;
			selectPos(depth + 1, i + 1, n, info);
			winPos[i] = false;
		}
	}
	
	public int calc(int[] info1, int[] info2) {
		int res = 0;
		
		for(int i = 0 ; i < 10 ; i++) {
			if(info1[i] > info2[i])
				res += 10 - i;
			else if(info2[i] != 0)
				res -= 10 - i;
		}
		
		return res;
	}
}

File: common/programmers/양궁대회/양궁대회_고진석.py
minArr = []
res = [0 for _ in range(11)]
maxScore = -100
N = 0

def solution(n, info):
    global minArr, N
    minArr = [x+1 for x in info]
    N = n

    winpos = []
    for i in range(10):
        winpos.append(i)
        dfs(winpos, i, info)
        winpos.pop(-1)
    
    if maxScore <= 0:
        return [-1]
    return res

def dfs(winpos, idx, info):
    sum = 0
    for p in winpos:
        sum += minArr[p]
    if sum > N:
        return
    else:
        info1 = [0 for _ in range(11)]
        for p in winpos:
            info1[p] = minArr[p]
        info1[10] += N - sum
        global maxScore, res
        sc = score(info1,info)
        if sc > maxScore:
            maxScore = sc
            res = info1
        elif sc == maxScore:
            for i in range(10,-1,-1):
                if info1[i] > res[i]:
                    res = info1
                    break
                elif info1[i] < res[i]:
                    break
    for i in range(idx+1,10):
        winpos.append(i)
        dfs(winpos, i, info)
        winpos.pop(-1)

def score(info1, info2):
    res = 0
    for i in range(10):
        if info1[i] > info2[i]:
            res += 10-i
        elif info2[i] != 0:
            res -= 10-i
    return res

print(solution(10, [0, 0, 0, 0, 0, 0, 0, 0, 3, 4, 3]))